scipy/stats/_morestats.py

```diff
@@ -3904,3 +3904,82 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate', *,
     if not normalize:
         res *= (high-low)/(2.*pi)  # [1] (2.3.14) w/ (2.3.7)
     return res
+
+
+def directionalmean(samples, *, axis=0, nan_policy='propagate'):
+    """
+    Computes the directional mean of a sample of vectors.
+
+    Serves as equivalent of the sample mean for directional data whose
+    magnitude is irrelevant, e. g. unit vectors. The vectorial data
+    whose directionals means are computed must be located in the last
+    axis.
+
+    Parameters
+    ----------
+    samples : array_like
+        Input array. Must be at least two-dimensional.
+    axis : int, optional
+        Axis along which directional means are computed. Default is 0.
+    nan_policy : {'propagate', 'raise', 'omit'}, optional
+        Defines how to handle when input contains nan. 'propagate' returns nan,
+        'raise' throws an error, 'omit' performs the calculations ignoring nan
+        values. Default is 'propagate'.
+    Returns
+    -------
+    directionalmean : ndarray
+        Directional mean.
+
+    Notes
+    -----
+    This uses a definition of directional mean from [1]_.
+    Essentially, the calculation is as follows.
+
+    .. code-block:: python
+
+        mean=samples.mean(axis=0)
+        directionalmean = mean/np.linalg.norm(mean)
+
+    This definition is applicable for *directional* data but not
+    for *axial* data. In practice, this means that if your data
+    are symmetric around the origin (`x` has the same meaning as
+    -`x`), this function should not be used.
+
+    References
+    ----------
+    .. [1] Mardia, Jupp. (2000). *Directional Statistics*
+       (p. 163). Wiley.
+
+    Examples
+    --------
+    >>> from scipy.stats import directionalmean
+    >>> data = np.array([[0.8660254, 0.5, 0.], [0.8660254, -0.5, 0.]])
+    >>> directionalmean(data)
+    array([1., 0., 0.])
+
+    The `regular` sample mean in contrast does not lie on the unit sphere.
+
+    >>> data.mean(axis=0)
+    array([0.8660254, 0., 0.])
+
+    """
+    samples = np.asarray(samples)
+    if nan_policy not in ['raise', 'propagate', 'omit']:
+        raise ValueError("nan_policty must be one of 'propagate', 'omit', "
+                         "'raise'.")
+    if samples.ndim < 2:
+        raise ValueError("samples must at least be two-dimensional. "
+                         f"Instead samples has shape: {samples.shape!r}")
+    samples = np.moveaxis(samples, axis, 0)
+    contains_nan = np.isnan(samples).any()
+    if contains_nan:
+        if nan_policy == 'propagate':
+            return np.nan
+        elif nan_policy == 'omit':
+            mean = np.nanmean(samples, axis=0)
```

Comment:
I'd recommend not adding `nan_policy` for now. At some point we can just throw the decorator (lightly modified) on here, and I know it will be all correct then. Right now, though, I don't think this is what you want because if there is a NaN in one component it should probably behave as if all components of the same observation were NaNs (and omit those components from the mean).

Reply:
Ok, will take that stuff out for now. Would it help you if I put the corresponding tests in some gist for reuse later on?

Reply:
Thanks.